src/merkletree.ts

```diff
@@ -0,0 +1,177 @@
+import { BytesLike, HexString, toHex, compare } from './bytes';
+
+import {
+  MultiProof,
+  makeMerkleTree,
+  isValidMerkleTree,
+  getProof,
+  getMultiProof,
+  processProof,
+  processMultiProof,
+  renderMerkleTree,
+} from './core';
+
+import { MerkleTreeOptions, defaultOptions } from './options';
+import { invariant } from './utils/errors';
+
+export type MerkleTreeData<T> = {
+  format: string;
+  tree: HexString[];
+  values: { value: T; treeIndex: number }[];
+};
+
+export interface MerkleTree<T> {
+  root: HexString;
+  render(): string;
+  entries(): Iterable<[number, T]>;
+  validate(): void;
+  leafLookup(leaf: T): number;
+  getProof(leaf: number | T): HexString[];
+  getMultiProof(leaves: (number | T)[]): MultiProof<HexString, T>;
+  verify(leaf: number | T, proof: HexString[]): boolean;
+  verifyMultiProof(multiproof: MultiProof<BytesLike, number | T>): boolean;
+}
```

Comment:
We could define dump() here as 
```suggestion
  dump(): MerkleTreeData<T>;
}
```

The benefit would be that it would trigger a typescript error if some some class that extends `MerkleTreeImpl<T>` forgets to include `dump`.